### ISSUE TITLE
Emit event for when a specific property changes

### ISSUE THREAD

**Issue by bitpshr** [dojo/widget-core#419]
**Enhancement**

It can be tedious to detect when particular properties change using a single listener attached with `@onPropertiesChanged`. It may be nice to add a new event and an associated decorator to attach a listener for when a specific property changes.

Package Version: latest

**Expected behavior:**

```ts
@onPropertyChanged('fooProperty')
protected onFooChange(evt: PropertyChangedEvent<this, ComboBoxProperties>) {
	// do something
}

@onPropertyChanged('barProperty')
@onPropertyChanged('bazProperty')
protected onFooChange(evt: PropertyChangedEvent<this, ComboBoxProperties>) {
	// do something
}

@onPropertyChanged('quxProperty')
protected onFooChange(evt: PropertyChangedEvent<this, ComboBoxProperties>) {
	// do something
}
```

**Actual behavior:**

```ts
@onPropertiesChanged
protected onPropertiesChanged(evt: PropertiesChangeEvent<this, ComboBoxProperties>) {
	if (includes(evt.changedPropertyKeys, 'fooProperty')) {
		// do something
	}
	else if (includes(evt.changedPropertyKeys, 'barProperty') || includes(evt.changedPropertyKeys, 'bazProperty')) {
		// do something
	}
	else if (includes(evt.changedPropertyKeys, 'quxProperty') {
		// do something
	}
}
```

**Comment by kitsonk** [comment]
Would it be better to have the `@onPropertiesChanged` decorator take an optional argument of `props` that is of type `string | string[]` which could then be used like:

```typescript
@onPropertiesChanged
protected onPropertiesChanged(evt: PropertiesChangeEvent<this, ComboBoxProperties>) {}

@onPropertiesChanged('fooProperty')
protected onFooChanged(evt: PropertiesChangeEvent<this, ComboBoxProperties>) {}

@onPropertiesChanged([ 'barProperty', 'bazProperty' ])
protected onBarBazChanged(evt: PropertiesChangeEvent<this, ComboBoxProperties>) {}
```

Also there is a question too if simple globbing could be allowed, like `*Property` or `foo*`, or even `*` as a substitute for no argument.

**Comment by agubler** [comment]
If we were to go with an overloaded decorator `onPropertiesChange` perhaps `function onPropertiesChange(...args: string[])` would be a better API so that the consumer could do (below), similar to how `this.classes` is used.

```ts
@onPropertiesChanged('fooProperty')
protected onFooChanged(evt: PropertiesChangeEvent<this, ComboBoxProperties>) {}

@onPropertiesChanged('barProperty', 'bazProperty')
protected onBarBazChanged(evt: PropertiesChangeEvent<this, ComboBoxProperties>) {}
```

I don't believe that globbing for this properties changes would be necessary.

**Comment by kitsonk** [comment]
Good point, if someone had an array of properties, they could just as easily:

```typescript
@onPropertiesChanged(...props)
```

**Comment by matt-gadd** [comment]
The initial thought on all of our decorators was just to support them as stackable rather than making them variable in rarity or take an array for simplicity. If we change this for `@onPropertiesChanged`, we should change also this for `@diffProperty` for consistency. That does raise questions on other stackable decorators like `@theme` also.

**Comment by kitsonk** [comment]
Not that it is a huge concern, but isn't stacking inefficient?  In this instance, the decorator could be much more efficient in installing a listener that checks a set of events, versus sort of having a myriad of listeners (or semi complex branching code demonstrated above).

**Comment by matt-gadd** [comment]
@kitsonk the simple answer is no, all the `onPropertiesChanged` functions are called off one event. the only cost is the registration of the decorator.

**Comment by bitpshr** [comment]
Regardless of the API, would passing multiple properties imply a logical AND or logical OR when listening for changes?

**Comment by agubler** [comment]
`onPropertiesChanged` functionality has been superseded by changes included in #586 - The fine grain control is possible using the diff property reactions.

```ts
@diffProperty('foo', auto)
@diffProperty('bar', reference)
protected myPropertyChangeReaction(previousProperties: { foo: string, bar: object }, newProperties: { foo: string, bar: object }) {
    // do something if either `foo` or `bar` are considered changed by the diff'ing strategies.
}
```
